File: python/distribute-repeating-integers.py
1.
class Solution:
    def canDistribute(self, nums: List[int], quantity: List[int]) -> bool:
        def dfs(start):
            if start==len(quantity): return True
            for k, v in cnt.items():
                if v>=quantity[start]:
                    cnt[k]-=quantity[start]
                    if dfs(start+1): return True
                    cnt[k]+=quantity[start]
            return False

        cnt=collections.Counter(nums)
        quantity.sort(reverse=True)
        return dfs(0)

2.
class Solution:
    def canDistribute(self, nums: List[int], quantity: List[int]) -> bool:
        @lru_cache(None)
        def dfs(start, mask):
            if mask==(1<<m)-1: return True
            if start==len(vals): return False
            mx=max(quantity[j] for j in range(m) if ((1<<j) & mask)==0)
            if vals[start]<mx: return False
            if dfs(start+1, mask): return True

            for i in range(1<<m):
                if (i | mask)!=i: continue # new selection is a superset of existing set
                sum=0
                for j in range(m):
                    if ((1<<j) & i) and ((1<<j) & mask)==0:
                        sum+=quantity[j]

                if sum<=vals[start]:
                    if dfs(start+1, i): return True

            return False

        m=len(quantity)
        cnt=collections.Counter(nums)
        vals=sorted(cnt.values(), reverse=True)
        return dfs(0, 0)

File: cpp/distribute-repeating-integers.cpp
1655. Distribute Repeating Integers
Hard

69

6

Add to List

Share
You are given an array of n integers, nums, where there are at most 50 unique values in the array. You are also given an array of m customer order quantities, quantity, where quantity[i] is the amount of integers the ith customer ordered. Determine if it is possible to distribute nums such that:

The ith customer gets exactly quantity[i] integers,
The integers the ith customer gets are all equal, and
Every customer is satisfied.
Return true if it is possible to distribute nums according to the above conditions.

 

Example 1:

Input: nums = [1,2,3,4], quantity = [2]
Output: false
Explanation: The 0th customer cannot be given two different integers.
Example 2:

Input: nums = [1,2,3,3], quantity = [2]
Output: true
Explanation: The 0th customer is given [3,3]. The integers [1,2] are not used.
Example 3:

Input: nums = [1,1,2,2], quantity = [2,2]
Output: true
Explanation: The 0th customer is given [1,1], and the 1st customer is given [2,2].
Example 4:

Input: nums = [1,1,2,3], quantity = [2,2]
Output: false
Explanation: Although the 0th customer could be given [1,1], the 1st customer cannot be satisfied.
Example 5:

Input: nums = [1,1,1,1,1], quantity = [2,3]
Output: true
Explanation: The 0th customer is given [1,1], and the 1st customer is given [1,1,1].
 

Constraints:

n == nums.length
1 <= n <= 105
1 <= nums[i] <= 1000
m == quantity.length
1 <= m <= 10
1 <= quantity[i] <= 105
There are at most 50 unique values in nums.

class Solution {
public:
    bool canDistribute(vector<int>& nums, vector<int>& quantity) {
        unordered_map<int, int> mp;
        for (auto& n:nums) mp[n]++;
        sort(quantity.begin(), quantity.end(), [](int& a, int& b){ return a>b; });
        return dfs(mp, quantity, 0);
    }
private:
    bool dfs(unordered_map<int, int>& mp, vector<int>& quantity, int start) {
        if (start==quantity.size()) return true;
        for (auto& iter:mp) {
            if (iter.second>=quantity[start]) {
                iter.second-=quantity[start];
                if (dfs(mp, quantity, start+1)) return true;
                iter.second+=quantity[start];
            }
        }
        return false;
    }
};

